instrumentation/github.com/aws/aws-lambda-go/otellambda/wrapLambdaHandler.go

```diff
@@ -0,0 +1,207 @@
+		return nil, e
+	}
+}
+
+// Ensure handler takes 0-2 values, with context
+// as its first value if two arguments exist
+func validateArguments(handler reflect.Type) (bool, error) {
+	handlerTakesContext := false
+	if handler.NumIn() > 2 {
+		return false, fmt.Errorf("handlers may not take more than two arguments, but handler takes %d", handler.NumIn())
+	} else if handler.NumIn() > 0 {
+		contextType := reflect.TypeOf((*context.Context)(nil)).Elem()
+		argumentType := handler.In(0)
+		handlerTakesContext = argumentType.Implements(contextType)
+		if handler.NumIn() > 1 && !handlerTakesContext {
+			return false, fmt.Errorf("handler takes two arguments, but the first is not Context. got %s", argumentType.Kind())
+		}
+	}
+
+	return handlerTakesContext, nil
+}
+
+// Ensure handler returns 0-2 values, with an error
+// as its first value if any exist
+func validateReturns(handler reflect.Type) error {
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+
+	switch n := handler.NumOut(); {
+	case n > 2:
+		return fmt.Errorf("handler may not return more than two values")
+	case n == 2:
+		if !handler.Out(1).Implements(errorType) {
+			return fmt.Errorf("handler returns two values, but the second does not implement error")
+		}
+	case n == 1:
+		if !handler.Out(0).Implements(errorType) {
+			return fmt.Errorf("handler returns a single value, but it does not implement error")
+		}
+	}
+
+	return nil
+}
+
+// Wraps and calls customer lambda handler then unpacks response as necessary
+func (whf *wrappedHandlerFunction) wrapperInternals(ctx context.Context, handlerFunc interface{}, eventJSON []byte, event reflect.Value, takesContext bool) (interface{}, error) {
+	wrappedLambdaHandler := reflect.ValueOf(whf.wrapper(handlerFunc))
+
+	argsWrapped := []reflect.Value{reflect.ValueOf(ctx), reflect.ValueOf(eventJSON), event, reflect.ValueOf(takesContext)}
+	response := wrappedLambdaHandler.Call(argsWrapped)[0].Interface().([]reflect.Value)
+
+	// convert return values into (interface{}, error)
+	var err error
+	if len(response) > 0 {
+		if errVal, ok := response[len(response)-1].Interface().(error); ok {
+			err = errVal
+		}
+	}
+	var val interface{}
+	if len(response) > 1 {
+		val = response[0].Interface()
+	}
+
+	return val, err
+}
+
+// WrapHandlerFunction Provides a lambda handler which wraps customer lambda handler with OTel Tracing
+func WrapHandlerFunction(handlerFunc interface{}, options ...Option) interface{} {
+	cfg := config{
+		TracerProvider:                 otel.GetTracerProvider(),
+		Flusher:                        &noopFlusher{},
+		EventToTextMapCarrierConverter: emptyEventToTextMapCarrierConverter,
+		Propagator:                     otel.GetTextMapPropagator(),
+	}
+	for _, opt := range options {
+		opt.apply(&cfg)
+	}
+
+	i := instrumentor{}
+	i.configuration = cfg
+	// Get a named tracer with package path as its name.
+	i.tracer = i.configuration.TracerProvider.Tracer(tracerName, trace.WithInstrumentationVersion(contrib.SemVersion()))
```

Comment:
This code block seems to be repeated.  Perhaps extract it as `newInstrumentor(opts ...Option) instrumentor`.